core/src/main/scala/org/apache/spark/sql/SnappySession.scala

```diff
@@ -1775,6 +1783,44 @@ object SnappySession extends Logging {
 
       allLiterals.foreach(_.collectedForPlanCaching = true)
     }
+
+    logDebug(s"qe.executedPlan = ${df.queryExecution.executedPlan}")
+
+    // This part is the defensive coding for all those cases where Tokenization
+    // support is not smart enough to deal with cases where the execution plan
+    // is modified in such a way that we cannot track those constants which
+    // need to be replaced in the subsequent execution
+    if (key != null) {
+      val nocaching = session.getContextObject[Boolean](
+        CachedPlanHelperExec.NOCACHING_KEY).getOrElse(false)
+      if (nocaching /* || allallbroadcastplans.nonEmpty */) {
+        logDebug(s"Invalidating the key because explicit nocaching")
+        key.invalidatePlan()
```

Comment:
Well it depends on when the broadcast variable is prepared. In some cases a broadcast variable is materialized quite early. A separate context and code generated happens for that. In some cases it is part of the top level plan.